File: ainpc-core-plugin/src/main/java/ro/ainpc/AINPCPlugin.java
package ro.ainpc;

import org.bukkit.configuration.file.FileConfiguration;
import org.bukkit.configuration.file.YamlConfiguration;
import org.bukkit.plugin.ServicePriority;
import org.bukkit.plugin.java.JavaPlugin;
import org.bukkit.command.PluginCommand;
import ro.ainpc.ai.DialogManager;
import ro.ainpc.ai.OpenAIService;
import ro.ainpc.ai.orchestration.AIOrchestrationService;
import ro.ainpc.api.AINPCPlatformApi;
import ro.ainpc.bootstrap.SchedulerCoordinator;
import ro.ainpc.commands.AINPCCommand;
import ro.ainpc.commands.AINPCTabCompleter;
import ro.ainpc.database.DatabaseManager;
import ro.ainpc.engine.DecisionEngine;
import ro.ainpc.engine.DialogueEngine;
import ro.ainpc.engine.FeaturePackLoader;
import ro.ainpc.engine.ScenarioEngine;
import ro.ainpc.gui.GuiService;
import ro.ainpc.listeners.ListenerRegistry;
import ro.ainpc.managers.ConversationSessionManager;
import ro.ainpc.managers.EmotionManager;
import ro.ainpc.managers.FamilyManager;
import ro.ainpc.managers.MemoryManager;
import ro.ainpc.managers.NPCManager;
import ro.ainpc.platform.AINPCPlatform;
import ro.ainpc.progression.ProgressionService;
import ro.ainpc.routine.RoutineService;
import ro.ainpc.spawn.HouseholdPersistenceService;
import ro.ainpc.spawn.NpcSpawnOrchestrator;
import ro.ainpc.story.StoryContextService;
import ro.ainpc.story.StoryStateService;
import ro.ainpc.utils.MessageUtils;
import ro.ainpc.world.NpcWorldBindingService;
import ro.ainpc.world.mapping.MappingWandService;

import java.io.File;
import java.util.logging.Level;

public class AINPCPlugin extends JavaPlugin {

    private static AINPCPlugin instance;
    
    private DatabaseManager databaseManager;
    private NPCManager npcManager;
    private MemoryManager memoryManager;
    private EmotionManager emotionManager;
    private FamilyManager familyManager;
    private ConversationSessionManager conversationSessionManager;
    private DialogManager dialogManager;
    private OpenAIService openAIService;
    private AIOrchestrationService aiOrchestrationService;
    private RoutineService routineService;
    private NpcSpawnOrchestrator npcSpawnOrchestrator;
    private HouseholdPersistenceService householdPersistenceService;
    private NpcWorldBindingService npcWorldBindingService;
    private MessageUtils messageUtils;
    private AINPCPlatform platform;
    private ListenerRegistry listenerRegistry;
    private SchedulerCoordinator schedulerCoordinator;
    private File questConfigFile;
    private FileConfiguration questConfig;
    
    // Motoare AI
    private DecisionEngine decisionEngine;
    private DialogueEngine dialogueEngine;
    private ScenarioEngine scenarioEngine;
    private FeaturePackLoader featurePackLoader;
    private ProgressionService progressionService;
    private StoryContextService storyContextService;
    private StoryStateService storyStateService;
    private GuiService guiService;
    private MappingWandService mappingWandService;

    @Override
    public void onEnable() {
        instance = this;
        
        // Salveaza configuratia default
        saveDefaultConfig();
        getConfig().options().copyDefaults(true);
        saveConfig();
        loadQuestConfig();
        
        // Initializeaza utilitarele
        messageUtils = new MessageUtils(this);

        // Initializeaza platforma core
        platform = new AINPCPlatform(this);
        platform.initialize();
        
        // Initializeaza baza de date
        getLogger().info("Initializare baza de date...");
        databaseManager = new DatabaseManager(this);
        if (!databaseManager.initialize()) {
            getLogger().severe("Nu s-a putut initializa baza de date! Pluginul se opreste.");
            getServer().getPluginManager().disablePlugin(this);
            return;
        }
        npcWorldBindingService = new NpcWorldBindingService(this);
        householdPersistenceService = new HouseholdPersistenceService(this);
        
        // Initializeaza serviciul OpenAI
        getLogger().info("Initializare serviciu OpenAI...");
        openAIService = new OpenAIService(this);
        aiOrchestrationService = new AIOrchestrationService(this);
        openAIService.runDiagnosticsAsync("startup");
        
        // Incarca Feature Packs
        getLogger().info("Incarcare Feature Packs...");
        featurePackLoader = new FeaturePackLoader(this);
        featurePackLoader.loadAllPacks();
        
        // Initializeaza managerii
        getLogger().info("Initializare manageri...");
        memoryManager = new MemoryManager(this);
        emotionManager = new EmotionManager(this);
        familyManager = new FamilyManager(this);
        npcManager = new NPCManager(this);
        routineService = new RoutineService(this);
        npcSpawnOrchestrator = new NpcSpawnOrchestrator(this);
        dialogManager = new DialogManager(this);
        conversationSessionManager = new ConversationSessionManager(this);
        
        // Incarca NPC-urile din baza de date
        npcManager.loadAllNPCs();
        npcManager.discoverExistingVillagers();
        npcManager.reconcileDuplicateLiveNPCEntities("startup");
        npcManager.restoreMissingNPCsInLoadedChunks();
        npcManager.enforceControlledEntitySettings("startup");
        int backfilledProfiles = npcManager.ensureAllNPCsHaveProfiles();
        int backfilledWorldBindings = npcManager.backfillWorldBindingsFromAnchors();
        getLogger().info("Profiluri NPC verificate. Profiluri create/backfill: " + backfilledProfiles);
        getLogger().info("Binding-uri NPC-world inferate/backfill: " + backfilledWorldBindings);
        
        // Initializeaza motoarele AI
        getLogger().info("Initializare motoare AI...");
        decisionEngine = new DecisionEngine(this);
        dialogueEngine = new DialogueEngine(this, openAIService);
        scenarioEngine = new ScenarioEngine(this);
        progressionService = new ProgressionService(this);
        storyStateService = new StoryStateService(this);
        storyContextService = new StoryContextService(this);
        guiService = new GuiService(this);
        mappingWandService = new MappingWandService(this);
        
        // Inregistreaza comenzile
        getLogger().info("Inregistrare comenzi...");
        AINPCCommand command = new AINPCCommand(this);
        PluginCommand ainpcCommand = getCommand("ainpc");
        if (ainpcCommand == null) {
            getLogger().severe("Comanda 'ainpc' nu a fost gasita in plugin.yml. Pluginul se opreste.");
            getServer().getPluginManager().disablePlugin(this);
            return;
        }
        ainpcCommand.setExecutor(command);
        ainpcCommand.setTabCompleter(new AINPCTabCompleter(this));

        PluginCommand npcQuestCommand = getCommand("npcquest");
        if (npcQuestCommand != null) {
            npcQuestCommand.setExecutor(command);
            npcQuestCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'npcquest' nu a fost gasita in plugin.yml.");
        }

        PluginCommand questCommand = getCommand("quest");
        if (questCommand != null) {
            questCommand.setExecutor(command);
            questCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'quest' nu a fost gasita in plugin.yml.");
        }

        PluginCommand progressionCommand = getCommand("progression");
        if (progressionCommand != null) {
            progressionCommand.setExecutor(command);
            progressionCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'progression' nu a fost gasita in plugin.yml.");
        }

        PluginCommand contractCommand = getCommand("contract");
        if (contractCommand != null) {
            contractCommand.setExecutor(command);
            contractCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'contract' nu a fost gasita in plugin.yml.");
        }

        PluginCommand dutyCommand = getCommand("duty");
        if (dutyCommand != null) {
            dutyCommand.setExecutor(command);
            dutyCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'duty' nu a fost gasita in plugin.yml.");
        }

        PluginCommand bountyCommand = getCommand("bounty");
        if (bountyCommand != null) {
            bountyCommand.setExecutor(command);
            bountyCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'bounty' nu a fost gasita in plugin.yml.");
        }

        PluginCommand eventCommand = getCommand("event");
        if (eventCommand != null) {
            eventCommand.setExecutor(command);
            eventCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'event' nu a fost gasita in plugin.yml.");
        }

        PluginCommand tutorialCommand = getCommand("tutorial");
        if (tutorialCommand != null) {
            tutorialCommand.setExecutor(command);
            tutorialCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'tutorial' nu a fost gasita in plugin.yml.");
        }

        PluginCommand ritualCommand = getCommand("ritual");
        if (ritualCommand != null) {
            ritualCommand.setExecutor(command);
            ritualCommand.setTabCompleter(new AINPCTabCompleter(this));
        } else {
            getLogger().warning("Comanda 'ritual' nu a fost gasita in plugin.yml.");
        }
        
        // Inregistreaza listenerele
        getLogger().info("Inregistrare listenere...");
        listenerRegistry = new ListenerRegistry(this);
        listenerRegistry.registerAll();

        schedulerCoordinator = new SchedulerCoordinator(this);
        schedulerCoordinator.start();
        getServer().getServicesManager().register(AINPCPlatformApi.class, platform, this, ServicePriority.Normal);
        
        getLogger().info("========================================");
        getLogger().info("AI NPC Plugin v" + getPluginMeta().getVersion() + " activat!");
        getLogger().info("NPC-uri incarcate: " + npcManager.getNPCCount());
        getLogger().info("Addonuri inregistrate: " + platform.getAddonRegistry().size());
        getLogger().info("World admin: " + platform.getWorldAdmin().getRegionCount() + " regiuni / "
            + platform.getWorldAdmin().getPlaceCount() + " places / "
            + platform.getWorldAdmin().getNodeCount() + " noduri");
        getLogger().info("========================================");
    }

    @Override
    public void onDisable() {
        if (scenarioEngine != null) {
            scenarioEngine.stopAllQuestTracking();
            getLogger().info("Salvare progres quest-uri...");
            scenarioEngine.flushQuestProgress();
        }

        // Salveaza toate datele NPC-urilor
        if (npcManager != null) {
            getLogger().info("Salvare date NPC-uri...");
            npcManager.saveAllNPCs();
        }
        
        // Inchide conexiunea la baza de date
        if (databaseManager != null) {
            getLogger().info("Inchidere conexiune baza de date...");
            databaseManager.close();
        }

        if (platform != null) {
            platform.shutdown();
        }
        if (guiService != null) {
            guiService.sessions().closeAll();
        }
        getServer().getServicesManager().unregisterAll(this);
        
        getLogger().info("AI NPC Plugin dezactivat!");
    }
    
    public void reload() {
        reloadConfig();
        getConfig().options().copyDefaults(true);
        saveConfig();
        loadQuestConfig();
        messageUtils = new MessageUtils(this);
        if (platform != null) {
            platform.reloadFromConfig();
        }
        openAIService = new OpenAIService(this);
        aiOrchestrationService = new AIOrchestrationService(this);
        openAIService.runDiagnosticsAsync("reload");
        if (memoryManager != null) {
            dialogueEngine = new DialogueEngine(this, openAIService);
        }
        reloadContent();
        getLogger().info("Configuratie reincarcata!");
    }

    public void reloadContent() {
        if (featurePackLoader != null) {
            featurePackLoader.loadAllPacks();
        }
        if (scenarioEngine != null) {
            scenarioEngine.reloadTemplates();
        }
        storyStateService = new StoryStateService(this);
        storyContextService = new StoryContextService(this);
        if (npcManager != null) {
            npcManager.ensureAllNPCsHaveProfiles();
        }
    }

    private void loadQuestConfig() {
        if (!getDataFolder().exists()) {
            getDataFolder().mkdirs();
        }

        questConfigFile = new File(getDataFolder(), "quests.yml");
        if (!questConfigFile.exists()) {
            saveResource("quests.yml", false);
        }

        questConfig = YamlConfiguration.loadConfiguration(questConfigFile);
    }

    // Getters
    public static AINPCPlugin getInstance() {
        return instance;
    }

    public DatabaseManager getDatabaseManager() {
        return databaseManager;
    }

    public NPCManager getNpcManager() {
        return npcManager;
    }

    public MemoryManager getMemoryManager() {
        return memoryManager;
    }

    public EmotionManager getEmotionManager() {
        return emotionManager;
    }

    public FamilyManager getFamilyManager() {
        return familyManager;
    }

    public ConversationSessionManager getConversationSessionManager() {
        return conversationSessionManager;
    }

    public AINPCPlatform getPlatform() {
        return platform;
    }

    public NpcSpawnOrchestrator getNpcSpawnOrchestrator() {
        return npcSpawnOrchestrator;
    }

    public NpcWorldBindingService getNpcWorldBindingService() {
        return npcWorldBindingService;
    }

    public HouseholdPersistenceService getHouseholdPersistenceService() {
        return householdPersistenceService;
    }

    public RoutineService getRoutineService() {
        return routineService;
    }

    public DialogManager getDialogManager() {
        return dialogManager;
    }

    public OpenAIService getOpenAIService() {
        return openAIService;
    }

    public AIOrchestrationService getAIOrchestrationService() {
        return aiOrchestrationService;
    }

    public MessageUtils getMessageUtils() {
        return messageUtils;
    }

    public DecisionEngine getDecisionEngine() {
        return decisionEngine;
    }

    public DialogueEngine getDialogueEngine() {
        return dialogueEngine;
    }

    public ScenarioEngine getScenarioEngine() {
        return scenarioEngine;
    }

    public ProgressionService getProgressionService() {
        return progressionService;
    }

    public StoryContextService getStoryContextService() {
        return storyContextService;
    }

    public StoryStateService getStoryStateService() {
        return storyStateService;
    }

    public FeaturePackLoader getFeaturePackLoader() {
        return featurePackLoader;
    }

    public GuiService getGuiService() {
        return guiService;
    }

    public MappingWandService getMappingWandService() {
        return mappingWandService;
    }

    public FileConfiguration getQuestConfig() {
        return questConfig;
    }
    
    public void debug(String message) {
        if (getConfig().getBoolean("debug")) {
            getLogger().log(Level.INFO, "[Debug] " + message);
        }
    }
}

File: ainpc-core-plugin/src/main/kotlin/ro/ainpc/AINPCPlugin.kt
package ro.ainpc

import org.bukkit.command.PluginCommand
import org.bukkit.plugin.java.JavaPlugin
import ro.ainpc.ai.DialogManager
import ro.ainpc.ai.OllamaService
import ro.ainpc.util.PermissionAudit
import ro.ainpc.ai.OpenAIService
import ro.ainpc.ai.RelationshipService
import ro.ainpc.ai.orchestration.AIOrchestrationService
import ro.ainpc.api.AINPCPlatformApi
import ro.ainpc.bootstrap.PackFileWatcher
import ro.ainpc.bootstrap.PerformanceMonitor
import ro.ainpc.bootstrap.SchedulerCoordinator
import ro.ainpc.bootstrap.ServiceRegistry
import ro.ainpc.database.DatabaseManager
import ro.ainpc.debug.RecentEventsBuffer
import ro.ainpc.economy.BankingService
import ro.ainpc.economy.EconomyService
import ro.ainpc.economy.NpcEconomyService
import ro.ainpc.economy.ShopService
import ro.ainpc.economy.VaultEconomyHook
import ro.ainpc.engine.DecisionEngine
import ro.ainpc.engine.DialogueEngine
import ro.ainpc.engine.FeaturePackLoader
import ro.ainpc.engine.QuestAuthoringService
import ro.ainpc.engine.ScenarioEngine
import ro.ainpc.environment.EnvironmentEngine
import ro.ainpc.environment.SeasonalBehaviorService
import ro.ainpc.gui.GuiService
import ro.ainpc.listeners.ListenerRegistry
import ro.ainpc.managers.ConversationSessionManager
import ro.ainpc.managers.EmotionManager
import ro.ainpc.managers.FamilyManager
import ro.ainpc.managers.MemoryManager
import ro.ainpc.managers.NPCManager
import ro.ainpc.mcp.McpRuntimeClient
import ro.ainpc.mcp.bridge.McpCommandQueue
import ro.ainpc.mcp.bridge.RuntimeSnapshotProducer
import ro.ainpc.platform.AINPCPlatform
import ro.ainpc.progression.PlayerProgressionService
import ro.ainpc.progression.ProgressionService
import ro.ainpc.reputation.ReputationService
import ro.ainpc.routine.RoutineCoordinator
import ro.ainpc.routine.RoutineService
import ro.ainpc.routine.SocialCoordinator
import ro.ainpc.spawn.AutoSettlementGenerator
import ro.ainpc.spawn.HouseholdPersistenceService
import ro.ainpc.spawn.NpcSpawnOrchestrator
import ro.ainpc.story.RandomWorldEventService
import ro.ainpc.story.StoryAuthoringService
import ro.ainpc.story.StoryContextService
import ro.ainpc.story.StoryReactionService
import ro.ainpc.story.StoryStateService
import ro.ainpc.utils.MessageUtils
import ro.ainpc.world.NpcWorldBindingService
import ro.ainpc.world.mapping.MappingWandService
import ro.ainpc.world.scan.VanillaVillageScanService
import java.io.File
import java.util.logging.Level

class AINPCPlugin : JavaPlugin() {
    val services = ServiceRegistry(this)

    val databaseManager: DatabaseManager get() = services.databaseManager
    val npcManager: NPCManager get() = services.npcManager
    val memoryManager: MemoryManager get() = services.memoryManager
    val emotionManager: EmotionManager get() = services.emotionManager
    val familyManager: FamilyManager get() = services.familyManager
    val conversationSessionManager: ConversationSessionManager get() = services.conversationSessionManager
    val dialogManager: DialogManager get() = services.dialogManager
    val openAIService: OpenAIService get() = services.openAIService
    val ollamaService: OllamaService get() = services.ollamaService
    val aiOrchestrationService: AIOrchestrationService get() = services.aiOrchestrationService
    val mcpRuntimeClient: McpRuntimeClient get() = services.mcpRuntimeClient
    val snapshotProducer: RuntimeSnapshotProducer? get() = services.snapshotProducer
    val mcpCommandQueue: McpCommandQueue get() = services.mcpCommandQueue
    val performanceMonitor: PerformanceMonitor get() = services.performanceMonitor
    val routineService: RoutineService get() = services.routineService
    val routineCoordinator: RoutineCoordinator get() = services.routineCoordinator
    val autoSettlementGenerator: AutoSettlementGenerator get() = services.autoSettlementGenerator
    val vanillaVillageScanService: VanillaVillageScanService get() = services.vanillaVillageScanService
    val npcSpawnOrchestrator: NpcSpawnOrchestrator get() = services.npcSpawnOrchestrator
    val householdPersistenceService: HouseholdPersistenceService get() = services.householdPersistenceService
    val npcWorldBindingService: NpcWorldBindingService get() = services.npcWorldBindingService
    val messageUtils: MessageUtils get() = services.messageUtils
    val platform: AINPCPlatform get() = services.platform
    val listenerRegistry: ListenerRegistry get() = services.listenerRegistry
    val packFileWatcher: PackFileWatcher get() = services.packFileWatcher
    val schedulerCoordinator: SchedulerCoordinator get() = services.schedulerCoordinator
    val questConfigFile: File get() = services.questConfigFile
    val questConfig: org.bukkit.configuration.file.FileConfiguration get() = services.questConfig
    val decisionEngine: DecisionEngine get() = services.decisionEngine
    val dialogueEngine: DialogueEngine get() = services.dialogueEngine
    val scenarioEngine: ScenarioEngine get() = services.scenarioEngine
    val featurePackLoader: FeaturePackLoader get() = services.featurePackLoader
    val progressionService: ProgressionService get() = services.progressionService
    val playerProgressionService: PlayerProgressionService get() = services.playerProgressionService
    val storyContextService: StoryContextService get() = services.storyContextService
    val storyAuthoringService: StoryAuthoringService get() = services.storyAuthoringService
    val storyReactionService: StoryReactionService get() = services.storyReactionService
    val randomWorldEventService: RandomWorldEventService get() = services.randomWorldEventService
    val storyStateService: StoryStateService get() = services.storyStateService
    val environmentEngine: EnvironmentEngine get() = services.environmentEngine
    val seasonalBehaviorService: SeasonalBehaviorService get() = services.seasonalBehaviorService
    val authoringService: QuestAuthoringService get() = services.authoringService
    val guiService: GuiService get() = services.guiService
    val mappingWandService: MappingWandService get() = services.mappingWandService
    val economyService: EconomyService get() = services.economyService
    val shopService: ShopService get() = services.shopService
    val npcEconomyService: NpcEconomyService get() = services.npcEconomyService
    val bankingService: BankingService get() = services.bankingService
    val vaultEconomyHook: VaultEconomyHook get() = services.vaultEconomyHook
    val relationshipService: RelationshipService get() = services.relationshipService
    val reputationService: ReputationService get() = services.reputationService
    val socialCoordinator: SocialCoordinator get() = services.socialCoordinator
    val recentEventsBuffer: RecentEventsBuffer get() = services.recentEventsBuffer

    override fun onEnable() {
        instance = this
        saveDefaultConfig()
        config.options().copyDefaults(true)
        saveConfig()
        services.loadQuestConfig()
        validateConfig()
        saveResourceIfAbsent("castel-world-admin.yml")
        saveResourceIfAbsent("settlements.yml")
        saveResourceIfAbsent("building_templates.yml")
        saveResourceIfAbsent("behavior_profiles.yml")
        val packsDir = File(dataFolder, "packs")
        if (!packsDir.exists()) {
            packsDir.mkdirs()
        }
        saveResourceIfAbsent("packs/compat-26-1-2.yml")

        val result = services.initialize()
        if (!result.success) {
            val error = result.error
            logger.severe("Initializarea pluginului a esuat: ${error?.message}")
            server.pluginManager.disablePlugin(this)
        }
    }

    override fun onDisable() {
        services.shutdown()
    }

    fun reload() {
        reloadConfig()
        config.options().copyDefaults(true)
        saveConfig()
        services.reload()
    }

    fun reloadContent() {
        services.reloadContent()
    }

    fun debug(message: String) {
        if (config.getBoolean("debug.enabled", false)) {
            logger.log(Level.INFO, "[Debug] $message")
        }
    }

    private fun saveResourceIfAbsent(path: String) {
        try {
            saveResource(path, false)
        } catch (ignored: Exception) {
            logger.fine("$path deja exista sau nu este disponibil.")
        }
    }

    private fun validateConfig() {
        val warnings = mutableListOf<String>()
        if (!config.contains("features.ai")) {
            warnings.add("config.yml: 'features.ai' nu este definit; se va folosi implicit false.")
        }
        if (!config.contains("openai.api_key") && config.getBoolean("features.ai", false)) {
            warnings.add("config.yml: 'openai.api_key' nu este definit, dar features.ai=true. AI-ul nu va functiona.")
            warnings.add("  >> Actiune: seteaza OPENAI_API_KEY in variabilele de mediu sau openai.api_key in config.yml.")
        }
        if (!config.contains("database.dialect")) {
            warnings.add("config.yml: 'database.dialect' nu este definit; se va folosi 'sqlite'.")
        }
        val aiEnabled = config.getBoolean("features.ai", false)
        val routineEnabled = config.getBoolean("routine.enabled", false)
        if (routineEnabled && !aiEnabled) {
            warnings.add("config.yml: routine.enabled=true dar features.ai=false. NPC-urile nu vor avea dialog AI.")
            warnings.add("  >> Actiune: seteaza features.ai=true sau dezactiveaza routine.enabled.")
        }
        if (warnings.isNotEmpty()) {
            logger.warning("=== Validare config.yml ===")
            for (w in warnings) {
                logger.warning("  ! $w")
            }
        }
        PermissionAudit(this).runStartupAudit()
        PermissionAudit.auditRegisteredCommands(this)
    }

    companion object {
        private lateinit var instance: AINPCPlugin

        @JvmStatic
        fun getInstance(): AINPCPlugin = instance
    }
}
